src/connection.rs

```diff
@@ -20,59 +20,297 @@
+
+            // Create a datachannel with label 'data'
+            let data_channel = connection
+                .create_data_channel(
+                    "data",
+                    Some(RTCDataChannelInit {
+                        negotiated: None,
+                        id: Some(channel_id),
+                        ordered: None,
+                        max_retransmits: None,
+                        max_packet_life_time: None,
+                        protocol: None,
+                    }),
+                )
+                .map_err(|e| io::Error::new(io::ErrorKind::Other, format!("webrtc error: {}", e)))
+                .await?;
+
+            // No need to hold the lock during the DTLS handshake.
+            drop(connection);
+
+            let (data_channel_rx, data_channel_tx) = oneshot::channel::<Arc<DetachedDataChannel>>();
+
+            // Wait until the data channel is opened and detach it.
+            data_channel
+                .on_open({
+                    let data_channel = data_channel.clone();
+                    Box::new(move || {
+                        debug!(
+                            "Data channel '{}'-'{}' open",
+                            data_channel.label(),
+                            data_channel.id()
+                        );
+
+                        Box::pin(async move {
+                            let data_channel = data_channel.clone();
+                            match data_channel.detach().await {
+                                Ok(detached) => {
+                                    if let Err(_) = data_channel_rx.send(detached) {
+                                        error!("data_channel_tx dropped");
+                                    }
+                                },
+                                Err(e) => {
+                                    error!("Can't detach data channel: {}", e);
+                                },
+                            };
+                        })
+                    })
+                })
+                .await;
+
+            // Wait until data channel is opened and ready to use
+            match data_channel_tx.await {
+                Ok(detached) => Ok(detached),
+                Err(e) => Err(io::Error::new(io::ErrorKind::Other, e.to_string())),
+            }
+        })
+    }
+
+    fn poll_outbound(
+        &self,
+        cx: &mut Context<'_>,
+        s: &mut Self::OutboundSubstream,
+    ) -> Poll<Result<Self::Substream, Self::Error>> {
+        match ready!(s.as_mut().poll(cx)) {
+            Ok(detached) => {
+                let ch = PollDataChannel::new(detached);
+
+                let mut channels = self.inner.data_channels.lock().unwrap();
+                channels.insert(ch.stream_identifier(), ch.clone());
+
+                Poll::Ready(Ok(ch))
+            },
+            Err(e) => Poll::Ready(Err(e)),
+        }
+    }
+
+    fn destroy_outbound(&self, _s: Self::OutboundSubstream) {
+        // noop
```

Comment:
One thing to point out is that the future of `_s` might have been polled a bit but not completely. In other words, the future might currently be waiting at one of the `await` points, and dropping the future will abruptly interrupt the execution. I don't really know whether it's a problem, though.

Reply:
I would suggest adding a comment explaining this in `open_outbound`, because even if it's fine now it's something that someone who might touch the code later needs to be aware of